backends/gstreamer/src/source.rs

```diff
@@ -38,113 +28,183 @@ mod imp {
+                    gst::DebugColorFlags::empty(),
+                    "Servo source",
+                ),
+                appsrc: app_src,
+                srcpad: ghost_pad,
+            }
+        }
+
+        // Adds interface implementations in the class
+        fn type_init(type_: &mut subclass::InitializingType<Self>) {
+            type_.add_interface::<gst::URIHandler>();
         }
 
-        fn class_init(klass: &mut BinClass) {
+        // Called exactly once before the first instantiation of an instance. This
+        // sets up any type-specific things, in this specific case it installs the
+        // properties so that GObject knows about their existence and they can be
+        // used on instances of our type
+        fn class_init(klass: &mut subclass::simple::ClassStruct<Self>) {
             klass.set_metadata(
                 "Servo Media Source",
                 "Source/Audio/Video",
                 "Feed player with media data",
                 "Servo developers",
             );
-        }
 
-        pub fn set_size(&self, size: i64) {
-            if self.appsrc.get_size() != -1 {
-                return;
-            }
-            self.appsrc.set_size(size);
+            let caps = gst::Caps::new_any();
+            let src_pad_template = gst::PadTemplate::new(
+                "src",
+                gst::PadDirection::Src,
+                gst::PadPresence::Always,
+                &caps,
+            )
+            .unwrap();
+            klass.add_pad_template(src_pad_template);
         }
-
-        inner_appsrc_proxy!(end_of_stream, gst::FlowReturn);
-        inner_appsrc_proxy!(get_current_level_bytes, u64);
-        inner_appsrc_proxy!(get_max_bytes, u64);
-        inner_appsrc_proxy!(push_buffer, buffer, gst::Buffer, gst::FlowReturn);
-        inner_appsrc_proxy!(set_callbacks, callbacks, AppSrcCallbacks, ());
-        inner_appsrc_proxy!(set_stream_type, type_, AppStreamType, ());
     }
 
-    impl ObjectImpl<Bin> for ServoSrc {
-        fn constructed(&self, bin: &Bin) {
-            bin.parent_constructed();
-
-            self.add_element(bin, &self.appsrc.clone().upcast());
-
-            let pad = self
+    // The ObjectImpl trait provides the setters/getters for GObject properties.
+    // Here we need to provide the values that are internally stored back to the
+    // caller, or store whatever new value the caller is providing.
+    //
+    // This maps between the GObject properties and our internal storage of the
+    // corresponding values of the properties.
+    impl ObjectImpl for ServoSrc {
+        glib_object_impl!();
+
+        // Called right after construction of a new instance
+        fn constructed(&self, obj: &glib::Object) {
+            // Call the parent class' ::constructed() implementation first
+            self.parent_constructed(obj);
+
+            let bin = obj.downcast_ref::<gst::Bin>().unwrap();
+            self.add_element(bin, &self.appsrc.clone().upcast())
```

Comment:
Why the clone and upcast? And this should be `bin.add_element(&self.appsrc)`, you want to call the public function and not the virtual method :)